contrib/babelfishpg_tsql/src/forjson.c

```diff
@@ -0,0 +1,163 @@
+}
+
+
+/*
+ * Map an SQL row to an JSON element, taking the row from the active
+ * SPI cursor.
+ */
+static void
+SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_value)
+{
+	int			i;
+	const char  *sep="";
+	bool 		isnull;
+
+	appendStringInfoChar(result,'{');
+	for (i = 1; i <= SPI_tuptable->tupdesc->natts; i++)
+	{
+		char        *colname;
+		Datum		colval;
+		
+		colname = SPI_fname(SPI_tuptable->tupdesc, i);
+		
+		if (!strcmp(colname,"\?column\?")) /* When column name or alias is not provided */
+		{
+			ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("Column expressions and data sources without names or aliases cannot be formatted as JSON text using FOR JSON clause. Add alias to the unnamed column or table")));
+		}
+		
+		colval = SPI_getbinval(SPI_tuptable->vals[rownum],
+							   SPI_tuptable->tupdesc,
+							   i,
+							   &isnull);	
+			
+		if (!isnull)
+		{
+			appendStringInfo(result,sep);
+			sep = ",";
+			appendStringInfo(result, "%s",
+							 tsql_json_build_object(CStringGetDatum(colname), 
+							 						colval,
+							 						SPI_gettypeid(SPI_tuptable->tupdesc,i),false)->data);
+		}
+		else if (include_null_value)
+		{
+			appendStringInfo(result,sep);
+			sep = ",";
+			appendStringInfo(result, "%s",
+							 tsql_json_build_object(CStringGetDatum(colname), 
+							 						colval,
+							 						SPI_gettypeid(SPI_tuptable->tupdesc,i),true)->data);
+		}
+	}
+	appendStringInfoChar(result,'}');
+	if (rownum != SPI_processed-1)
+	{
+		appendStringInfoString(result,",");
+	}
+}
+
+/*
+ * Perform the operation based on the mode and directives in the input query
+ */
+static StringInfo
+tsql_query_to_json_internal(const char *query, int mode, bool include_null_value,
+				bool without_array_wrapper, bool root_name_present, const char *root_name)
+{
+	StringInfo	result;
+	uint64		i;
+
+	set_config_option("babelfishpg_tsql.sql_dialect", "tsql",
+					  (superuser() ? PGC_SUSET : PGC_USERSET),
+					  PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+	result = makeStringInfo();
+
+	SPI_connect();
+	if (SPI_execute(query, true, 0) != SPI_OK_SELECT)
+		ereport(ERROR,
+				(errcode(ERRCODE_DATA_EXCEPTION),
+				 errmsg("invalid query")));
+
+	/* ROOT option and WITHOUT_ARRAY_WRAPPER option cannot be used together in FOR JSON */
+	if (root_name_present && without_array_wrapper)
+	{
+		ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+					 errmsg("ROOT option and WITHOUT_ARRAY_WRAPPER option cannot be used together in FOR JSON. Remove one of these options")));
+	}
+
+	/* If root_name_present is TRUE then WITHOUT_ARRAY_WRAPPER will be FALSE */
+	if(root_name_present)
+		appendStringInfo(result, "{\"%s\":[",root_name);
+	else if (!without_array_wrapper)
+		appendStringInfoChar(result,'[');
+
+	/* Format the query result according to the mode specified by the query */
+	switch (mode)
+	{
+		case TSQL_FORJSON_AUTO: /* FOR JSON AUTO */
+			ereport(ERROR,
```

Comment:
Please add a test case if this code is reachable, otherwise use `assert` or `internal_error`.

Reply:
Are you sure this code is reachable? There is no error message, `AUTO mode is not supported`, in the test file.

Reply:
Have made changes as discussed.